featureflags/services/auth.py

```diff
@@ -1,233 +1,316 @@
+                    and auth_session_row.expiration_time > datetime.utcnow()
                 ):
-                    state = ExpiredAccessToken(
-                        row.auth_user, secret, session_key, row.expiration_time
+                    state = ExpiredAccessTokenState(
+                        user=auth_session_row.auth_user,
+                        secret=secret,
+                        ident=payload["session"],
+                        session_exp=auth_session_row.expiration_time,
                     )
                 else:
-                    state = SignedOutSession(session_key, secret)
-            else:
-                state = SignedOutSession(session_key, secret)
-        except jwt.InvalidSignatureError:  # if secret key was changed
-            payload = jwt.decode(
-                access_token, secret,
-                options={
-                    'verify_signature': False,
-                },
-                algorithms=["HS256"]
-            )
-            session_key = payload["session"]
-            state = SignedOutSession(session_key, secret)
+                    state = SignedOutState(
+                        user=None,
+                        ident=payload["session"],
+                        secret=secret,
+                    )
+
         else:
-            session_key = payload["session"]
-            if "user" in payload:
-                try:
-                    user = UUID(hex=payload["user"])
-                except ValueError:
-                    # FIXME: backward compatibility
-                    state = SignedOutSession(session_key, secret)
-                else:
-                    state = ValidAccessToken(user=user)
+            state = SignedOutState(
+                user=None,
+                ident=payload["session"],
+                secret=secret,
+            )
+
+    except jwt.InvalidSignatureError:  # if secret key was changed
+        payload = decode_jwt_token(secret, access_token, verify_signature=False)
+        state = SignedOutState(
+            user=None,
+            ident=payload["session"],
+            secret=secret,
+        )
+
+    else:
+        if "user" in payload:
+            try:
+                user = UUID(hex=payload["user"])
+            except ValueError:
+                # FIXME: backward compatibility
+                state = SignedOutState(
+                    user=None,
+                    ident=payload["session"],
+                    secret=secret,
+                )
             else:
-                state = EmptyAccessToken()
-    return Session(session_key, state, secret=secret)
+                state = ValidAccessTokenState(user=user)
+        else:
+            state = EmptyAccessTokenState(user=None)
+
+    return UserSession(ident=payload["session"], state=state, secret=secret)
+
+
+async def set_user_session_from_cookie(
+    request: Request, engine: aiopg.sa.Engine
+) -> None:
+    access_token = request.cookies.get(COOKIE_ACCESS_TOKEN)
```

Comment:
what if there is no cookie ? do we need to handle that case here ?

Reply:
that is handled in `create_user_session` in set_user_session_from_cookie (a few lines below, from 281 to 286) when setting `user_session`